Create tutorial for compound assignment operators (#48476)
dotnet/docs: docs/csharp/whats-new/tutorials/compound-assignment-operators.md

With traditional operators, each operation creates a new `GateAttendance` instance due to the immutable nature of records, leading to significant memory allocations.

When you run the simulation, you see detailed output showing:

- Gate-by-gate attendance numbers during different arrival periods.
- Total attendance tracking across all gates.
- A final comprehensive report with attendance statistics.

You can see a portion of the output:

```txt
Peak arrival time - all gates busy...

Peak rush period completed - all gates processed heavy traffic.

--- Gate Status After Main Rush (7:15 PM) ---
Main Floor Gates:
  Main-Floor-Gate-1: 145 attendees
  Main-Floor-Gate-2: 168 attendees
  Main-Floor-Gate-3: 149 attendees
  Main-Floor-Gate-4:  71 attendees
  Main Floor Subtotal: 533 attendees

Balcony Gates:
  Balcony-Gate-Left: 164 attendees
  Balcony-Gate-Right: 134 attendees
  Balcony Subtotal: 298 attendees

Total Current Attendance: 831 / 1000

--- Late Arrivals (7:15 PM - 7:30 PM) ---
Final patrons arriving before curtain...

Final arrivals processed - concert about to begin!
```

> This realistic scenario provides an excellent test case for measuring the performance benefits of compound assignment operators, as the frequent count updates mirror common patterns in real applications where objects are repeatedly modified.

Reviewer: This paragraph could maybe be left out.

Examine the starter `GateAttendance` record class:

:::code language="csharp" source="./snippets/CompoundAssignment/GateAttendance.cs" id="GateAttendanceStarter":::

> The `InitialImplementation.GateAttendance` record demonstrates the traditional approach to operator overloading in C#. Notice how both the increment operator (`++`) and addition operator (`+`) create entirely new instances of `GateAttendance` using the `with` expression. Each time you write `gate++` or `gate += partySize`, the operators allocate a new record instance with the updated `Count` value, then return that new instance. While this approach maintains immutability and thread safety, it comes at the cost of frequent memory allocations. In scenarios with many operations—like our concert simulation with hundreds of attendance updates—these allocations accumulate quickly, potentially impacting performance and increasing garbage collection pressure.

Reviewer: ```suggestion
The `InitialImplementation.GateAttendance` record demonstrates the traditional approach to operator overloading in C#. Notice how both the increment operator (`++`) and addition operator (`+`) create entirely new instances of `GateAttendance` using the `with` expression. Each time you write `gate++` or `gate += partySize`, the operators allocate a new record instance with the updated `Count` value, then return that new instance. While this approach maintains immutability and thread safety, it comes at the cost of frequent memory allocations. In scenarios with many operations—like the concert simulation with hundreds of attendance updates—these allocations accumulate quickly, potentially impacting performance and increasing garbage collection pressure.
```

## Implement compound assignment operators

C# 14 introduces user-defined compound assignment operators that enable in-place mutations instead of creating new instances. These operators provide a more efficient alternative to the traditional pattern while maintaining the familiar compound assignment syntax.

Compound assignment operators use a new syntax that declares `void` return methods with the `operator` keyword. Add the following operators to the `GateAttendance` class:

```csharp
public void operator +=(int value) => this.property += value;
public void operator ++() => this.property++;
```

The key differences from traditional operators are:

> - **Mutation**: They modify the current instance directly using `this`
> - **No new instances**: Unlike traditional operators that return new objects, compound operators modify existing ones
> - **Return type**: Compound assignment operators return `void`, not the type itself

Reviewer: ```suggestion
- **Mutation**: They modify the current instance directly using `this`.
- **No new instances**: Unlike traditional operators that return new objects, compound operators modify existing ones.
- **Return type**: Compound assignment operators return `void`, not the type itself.
```

When the compiler encounters compound assignment expressions like `a += b` or `++a`, it follows this resolution order:

> 1. **Check for compound assignment operator**: If the type defines a user-defined compound assignment operator (`+=`, `++`, etc.), use it directly
> 2. **Fallback to traditional expansion**: If no compound operator exists, expand to the traditional form (`a = a + b`)

Reviewer: ```suggestion
1. **Check for compound assignment operator**: If the type defines a user-defined compound assignment operator (for example, `+=` or `++`), use it directly.
2. **Fallback to traditional expansion**: If no compound operator exists, expand to the traditional form (`a = a + b`).
```

This means you can implement both approaches simultaneously. The compound operators take precedence when available, but the traditional operators serve as fallbacks for scenarios where compound assignment isn't suitable.

Compound assignment operators provide several advantages:

> - **Reduced allocations**: Modify objects in-place instead of creating new instances
> - **Improved performance**: Eliminate temporary object creation and reduce garbage collection pressure  
> - **Familiar syntax**: Use the same `+=`, `++` syntax developers already know
> - **Backward compatibility**: Traditional operators continue to work as fallbacks

Reviewer: ```suggestion
- **Reduced allocations**: Modify objects in-place instead of creating new instances.
- **Improved performance**: Eliminate temporary object creation and reduce garbage collection pressure.
- **Familiar syntax**: Use the same `+=`, `++` syntax developers already know.
- **Backward compatibility**: Traditional operators continue to work as fallbacks.
```

The new compound assignment operators are shown in the following code:

:::code language="csharp" source="./snippets/CompoundAssignment/GateAttendance.cs" id="CompoundAssignmentOperators":::

## Analyze finished sample

Now that you implemented the compound assignment operators, it's time to measure the performance improvement. To see the dramatic difference in memory allocations, run the [.NET Object Allocation tracking tool](/visualstudio/profiling/dotnet-alloc-tool) again on the updated code.

When you profile the application with the compound assignment operators enabled, you observe a remarkable reduction: only **10 `GateAttendance` objects** are allocated during the entire concert simulation, compared to the previous 134 allocations. This update represents a 92% reduction in object allocations!

The remaining 10 allocations come from the initial creation of the `GateAttendance` instances for each theater gate (four main floor gates + two balcony gates = six initial instances), plus a few more allocations from other parts of the simulation that don't use the compound operators.

This allocation reduction translates to real performance benefits:

> - **Reduced memory pressure**: Less frequent garbage collection cycles
> - **Better cache locality**: Fewer object creations mean less memory fragmentation  
> - **Improved throughput**: CPU cycles saved from allocation and collection overhead
> - **Scalability**: Benefits multiply in scenarios with higher operation volumes

Reviewer: ```suggestion
- **Reduced memory pressure**: Less frequent garbage collection cycles.
- **Better cache locality**: Fewer object creations mean less memory fragmentation.
- **Improved throughput**: CPU cycles saved from allocation and collection overhead.
- **Scalability**: Benefits multiply in scenarios with higher operation volumes.
```

The performance improvement becomes even more significant in production applications where similar patterns occur at much larger scales—imagine tracking millions of transactions, updating thousands of counters, or processing high-frequency data streams.

> Try identifying other opportunities for compound assignment operators in the codebase. Look for patterns where you see traditional assignment operations like `gates.MainFloorGates[1] = gates.MainFloorGates[1] + 4` and consider whether they could benefit from compound assignment syntax. While some of these operations are already using `+=` in the simulation code, the principle applies to any scenario where you're repeatedly modifying objects rather than creating new instances.

Reviewer: ```suggestion
Try identifying other opportunities for compound assignment operators in the codebase. Look for patterns where you see traditional assignment operations like `gates.MainFloorGates[1] = gates.MainFloorGates[1] + 4` and consider whether they could benefit from compound assignment syntax. While some of these operations are already using `+=` in the simulation code, the principle applies to any scenario where you repeatedly modify objects rather than creating new instances.
```
